modules/3d/src/hal_replacement.hpp

```diff
@@ -0,0 +1,188 @@
+    // tangential distortion coefficients
+    float p[2];
+    // amount of tangential distortion coefficients passed
+    int amt_p;
+    // prism distortion coefficients
+    float s[4];
+    // amount of prism distortion coefficients passed
+    int amt_s;
+    // tilt distortion coefficients
+    float tau_x, tau_y;
+    // to use tilt distortion coefficients or not
+    bool use_tau;
+};
+
+/**
+   @brief Project points from 3D world space to 2D screen space using rotation and translation matrix and camera intrinsic parameters
+   @param src_data Pointer to 3D points array with coordinates interleaved as X, Y, Z, X, Y, Z,..
+   @param src_step Step between consecutive 3D points
+   @param src_size Amount of points
+   @param dst_data Pointer to resulting projected 2D points with coordinates interleaved as u, v, u, v,..
+   @param dst_step Step between consecutive projected 2D points
+   @param rt_data Pointer to 3x4 array containing rotation-then-translation matrix
+   @param intr_data Pointer to camera intrinsics structure
+*/
+inline int hal_ni_project_points_pinhole32f(const float* src_data, size_t src_step, size_t src_size,
+                                            float* dst_data, size_t dst_step, const float* rt_data,
+                                            const cv_camera_intrinsics_pinhole_32f* intr_data)
+{ return CV_HAL_ERROR_NOT_IMPLEMENTED; }
+
+//! @cond IGNORED
+#define cv_hal_project_points_pinhole32f hal_ni_project_points_pinhole32f
+//! @endcond
+
+/**
+ * @brief Camera intrinsics structure, see projectPoints() documentation for details
+ */
+struct cv_camera_intrinsics_pinhole_64f
+{
+    // focal length, principal point
+    double fx, fy, cx, cy;
+    // radial distortion coefficients
+    double k[6];
+    // amount of radial distortion coefficients passed
+    int amt_k;
+    // tangential distortion coefficients
+    double p[2];
+    // amount of tangential distortion coefficients passed
+    int amt_p;
+    // prism distortion coefficients
+    double s[4];
+    // amount of prism distortion coefficients passed
+    int amt_s;
+    // tilt distortion coefficients
+    double tau_x, tau_y;
+    // to use tilt distortion coefficients or not
+    bool use_tau;
+};
+
+/**
+   @brief Project points from 3D world space to 2D screen space using rotation and translation matrix and camera intrinsic parameters
+   @param src_data Pointer to 3D points array with coordinates interleaved as X, Y, Z, X, Y, Z,..
+   @param src_step Step between consecutive 3D points
+   @param src_size Amount of points
+   @param dst_data Pointer to resulting projected 2D points with coordinates interleaved as u, v, u, v,..
+   @param dst_step Step between consecutive projected 2D points
+   @param rt_data Pointer to 3x4 array containing rotation-then-translation matrix
+   @param intr_data Pointer to camera intrinsics structure
+*/
+inline int hal_ni_project_points_pinhole64f(const double* src_data, size_t src_step, size_t src_size,
```

Comment:
The integration part of HAL is missing in the PR. `undistortPoints` was reworked in 5.x. I'll attach solution soon.